core/services/workflows/store/store_db.go

```diff
@@ -0,0 +1,362 @@
+	return d.transact(ctx, func(db *DBStore) error {
+		wex := &workflowExecutionRow{
+			ID:         state.ExecutionID,
+			WorkflowID: state.WorkflowID,
+			Status:     state.Status,
+		}
+		err := db.insertWorkflowExecution(ctx, wex)
+		if err != nil {
+			return fmt.Errorf("could not insert workflow execution %s: %w", state.ExecutionID, err)
+		}
+
+		ws := []workflowStepRow{}
+		for _, step := range state.Steps {
+			step, err := stateToStep(step)
+			if err != nil {
+				return err
+			}
+			ws = append(ws, step)
+		}
+		if len(ws) > 0 {
+			return db.upsertSteps(ctx, ws)
+		}
+		return nil
+	})
+}
+
+func (d *DBStore) upsertSteps(ctx context.Context, steps []workflowStepRow) error {
+	sql := `
+	INSERT INTO
+	workflow_steps(workflow_execution_id, ref, status, inputs, output_err, output_value, created_at)
+	VALUES (:workflow_execution_id, :ref, :status, :inputs, :output_err, :output_value, NOW())
+	ON CONFLICT ON CONSTRAINT uniq_workflow_execution_id_ref
+	DO UPDATE SET
+		workflow_execution_id = EXCLUDED.workflow_execution_id,
+		ref = EXCLUDED.ref,
+		status = EXCLUDED.status,
+		inputs = EXCLUDED.inputs,
+		output_err = EXCLUDED.output_err,
+		output_value = EXCLUDED.output_value,
+		updated_at = NOW();
+	`
+	stmt, args, err := sqlx.Named(sql, steps)
+	if err != nil {
+		return err
+	}
+	stmt = d.db.Rebind(stmt)
+	_, err = d.db.ExecContext(ctx, stmt, args...)
+	return err
+}
+
+func (d *DBStore) insertWorkflowExecution(ctx context.Context, execution *workflowExecutionRow) error {
+	sql := `
+	INSERT INTO
+	workflow_executions(id, workflow_id, status, created_at)
+	VALUES ($1, $2, $3, NOW())
+	`
+	_, err := d.db.ExecContext(ctx, sql, execution.ID, execution.WorkflowID, execution.Status)
+	return err
+}
+
+func (d *DBStore) transact(ctx context.Context, fn func(*DBStore) error) error {
+	return sqlutil.Transact(
+		ctx,
+		func(ds sqlutil.DataSource) *DBStore {
+			return &DBStore{db: ds}
+		},
+		d.db,
+		nil,
+		fn,
+	)
+}
+
+func (d *DBStore) GetUnfinished(ctx context.Context, offset, limit int) ([]WorkflowExecution, error) {
+	sql := `
+	SELECT
+		workflow_steps.workflow_execution_id AS ws_workflow_execution_id,
+		workflow_steps.ref AS ws_ref,
+		workflow_steps.status AS ws_status,
+		workflow_steps.inputs AS ws_inputs,
+		workflow_steps.output_err AS ws_output_err,
+		workflow_steps.output_value AS ws_output_value,
+		workflow_steps.created_at AS ws_created_at,
+		workflow_steps.updated_at AS ws_updated_at,
+		workflow_executions.id AS we_id,
+		workflow_executions.workflow_id AS we_workflow_id,
+		workflow_executions.status AS we_status,
+		workflow_executions.created_at AS we_created_at,
+		workflow_executions.updated_at AS we_updated_at,
+		workflow_executions.completed_at AS we_completed_at
+	FROM workflow_executions
+	JOIN workflow_steps
+	ON  workflow_steps.workflow_execution_id = workflow_executions.id
+	WHERE workflow_executions.status = $1
+	LIMIT $2
+	OFFSET $3
```

Comment:
>As a related note, I'm curious if we need to worry about the "Thundering Herd Problem" occuring from starting a large number of workflow executions at the same time. I could see this happening if a node has been disconnected from the network for awhile, causing many pending workflow executions, then reconnects.

I don't think so: we have a fixed size worker pool (currently set to a low value of 100), which means we'll only enqueue a maximum of 100 steps, with the remaining being throttled because the tasks get enqueued by putting them on a channel (so we'll always have at most 100 steps, with one being added every time one completes).

We also have an execution-level timeout which should ensure that at most we'll be re-executing the last 10 minutes worth of executions. (So currently, in staging, that would be 10 executions)